Bundle todo

mhevery opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?
```
[ ] Yes
[ ] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

mhevery committed:
docs(ivy): update outstanding work for todo

mhevery committed:
fix(ivy): named listener method for easier debugging.

mhevery committed:
feat(ivy): Add `outputs` support for `defineDirective` / `defineComponent`

mhevery committed:
feat(ivy): Bind to right context in nested template statmentes

Given
```
<div *ngFor=”…” (click)=“doSomething()”>
```

Before `doSomething` would execute on the inner template context, which
is incorrect. The correct behavior is to execute on the top level context
of the component.

pkozlowski-opensource commented on packages/compiler/src/render3/r3_view_compiler.ts:
```diff
@@ -71,10 +71,14 @@ export function compileDirective(
 
   // e.g. `attributes: ['role', 'listbox']`
   field('attributes', createHostAttributesArray(directive, outputCtx));
+  console.error('=======', directive);
```
`console.error`

mhevery replied:
fixed

pkozlowski-opensource commented on packages/compiler/test/render3/r3_view_compiler_template_spec.ts:
```diff
@@ -0,0 +1,97 @@
+/* These tests are codified version of the tests in compiler_canonical_spec.ts. Every
+  * test in compiler_canonical_spec.ts should have a corresponding test here.
+  */
+fdescribe('compiler compliance: listen()', () => {
```
`fdescribe`

mhevery replied:
fixed

pkozlowski-opensource commented on packages/core/test/bundling/todo/OUTSTANDING_WORK.md:
```diff
@@ -1,9 +1,31 @@
+ 
 ## Ivy Runtime
 - [ ] Work on `ViewContainerRef` needs to cause change detection so that `todo` app renders correctly on first render.
 - [ ] The todo input value box is not correctly rendering to checked for completed tasks.
```
This should be fixed by https://github.com/angular/angular/pull/23161, right?

mhevery replied:
updated

mhevery committed:
build: create a more focused build file for packages/compiler/render3

mhevery committed:
test(ivy): cleanup todo test app

vicb commented on packages/core/src/render3/instructions.ts:
```diff
@@ -1765,7 +1765,7 @@ export function wrapListenerWithDirtyLogic(view: LView, listenerFn: (e?: any) =>
 export function wrapListenerWithDirtyAndDefault(
     view: LView, listenerFn: (e?: any) => any): EventListener {
-  return function(e: Event) {
+  return function wrapListenerIn_markViewDirty(e: Event) {
```
follow CS `wrapListenerInMarkViewDirty` ?

mhevery replied:
I would like to keep this as is unless you feel strongly. Rest of the generated code follows this convention such as `ToDoAppComponent_NgIf_NgForOf_NgIf_Template_6`

vicb commented on packages/compiler/src/render3/r3_view_compiler.ts:
```diff
@@ -1055,7 +1061,15 @@ function createHostBindingsFunction(
+  return null;
+}
+
+function createOutputsObject(
```
factorize `createInputsObject` / `createOutputsObject` ?

mhevery replied:
refactored

vicb commented on packages/compiler/test/render3/r3_view_compiler_input_outputs_spec.ts:
```diff
@@ -0,0 +1,89 @@
+              })
+              export class MyComponent {
+                @Input() name;
+                @Input('renamed') original;
```
renamedInput ?

mhevery replied:
cleaned up

vicb commented on packages/compiler/test/render3/r3_view_compiler_input_outputs_spec.ts:
```diff
@@ -0,0 +1,89 @@
+                selector: '[my-directive]',
+              })
+              export class MyDirective {
+                @Input() dirName;
```
same comment as for the cmp: consistency for I vs O names

mhevery replied:
done

vicb commented on packages/compiler/test/render3/r3_view_compiler_input_outputs_spec.ts:
```diff
@@ -0,0 +1,89 @@
+import {compile, expectEmit} from './mock_compile';
+
+/* These tests are codified version of the tests in compiler_canonical_spec.ts. Every
+  * test in compiler_canonical_spec.ts should have a corresponding test here.
```
is it the case ?

mhevery replied:
removed comment

vicb commented:
Commit message typo -> "statmentes" (es) -> "statements" (en)

vicb reviewed (changes requested):
See inline comments in each commits

mhevery committed:
fix(ivy): correctly bind to component context

When compiling templates the compiler would often bind to
closest context rather than the component context.

The only time one should be binding to the cont component is
in explicit cases where the inner template declares local variable.

kara commented on packages/compiler/test/render3/r3_view_compiler_template_spec.ts:
```diff
@@ -0,0 +1,97 @@
+              if (cm) {
+                i0.ɵE(0,'div');
+                i0.ɵL('click',function MyComponent_NgForOf_NgForOf_NgForOf_Template_1_div_click_listener($event:any){
+                  const pd_b:any = ((<any>ctx.onClick($_r0$,$_r1$,$_r2$)) !== false);
```
This generated code doesn't match our canonical spec for listeners [here](https://github.com/angular/angular/blob/master/packages/core/test/render3/compiler_canonical/elements_spec.ts#L110). Is there a reason for that or is this just work that hasn't been done yet in the compiler?

mhevery replied:
We should convert the canonical specs to this format. Out of scope for this PR.

mhevery replied:
I have updated the generated code to match spec. It is a bit of a hack so please let take a look if you are OK with it.

mhevery committed:
docs(ivy): update STATUS.md with outstanding work

mhevery committed:
test(ivy): update todo app to http://todomvc.com

mhevery commented on packages/compiler/test/render3/r3_view_compiler_template_spec.ts:
```diff
@@ -0,0 +1,97 @@
+import {MockDirectory, setup} from '../aot/test_util';
+import {compile, expectEmit} from './mock_compile';
+
+/* These tests are codified version of the tests in compiler_canonical_spec.ts. Every
```
removed

kara reviewed (approved):
LGTM

mary-poppins commented:
You can preview 4560326 at https://pr23168-4560326.ngbuilds.io/.
You can preview 182c47f at https://pr23168-182c47f.ngbuilds.io/.
You can preview d64fa70 at https://pr23168-d64fa70.ngbuilds.io/.
You can preview 8ab9ba1 at https://pr23168-8ab9ba1.ngbuilds.io/.

mary-poppins commented:
You can preview 4579242 at https://pr23168-4579242.ngbuilds.io/.

vicb commented on packages/compiler/test/render3/r3_view_compiler_input_outputs_spec.ts:
```diff
@@ -0,0 +1,86 @@
+
+    // The template should look like this (where IDENT is a wild card for an identifier):
+    const template = `
+      static ngComponentDef = i0.ɵdefineComponent({
```
`i0` should not be hardcoded

mhevery replied:
change to IDENT

vicb commented on packages/compiler/test/render3/r3_view_compiler_input_outputs_spec.ts:
```diff
@@ -0,0 +1,86 @@
+      }
+    };
+
+    // The template should look like this (where IDENT is a wild card for an identifier):
```
there is no IDENT

mhevery replied:
now there is

vicb commented on packages/compiler/test/render3/r3_view_compiler_input_outputs_spec.ts:
```diff
@@ -0,0 +1,86 @@
+/**
```
looks like a conformance test ? should it be moved to that folder ?

mhevery replied:
Why do you think it is conformance. I think of it as unit test for the `r3_view_compiler.ts` but that test file is huge so i suffixed it by type `intpu_outputs`.

vicb replied:
> Why do you think it is conformance

Because the only assertion you have here is about the compiler output.
Probably we should have a clear definition of the types of test and what they should do.

mhevery replied:
discussed in person

vicb commented on packages/core/test/bundling/todo/index.ts:
```diff
@@ -6,73 +6,150 @@
+
+  private _title: String;
+  get title() { return this._title; }
+  set title(value: String) { this._title = value.trim(); }
```
string (multiple)

mhevery replied:
fixed

vicb commented on packages/core/test/bundling/todo/index.ts:
```diff
@@ -6,73 +6,150 @@
 
-  static DEFAULT_TODO = {text: '', done: false};
-}
+  private getWithCompleted(completed: Boolean) {
```
boolean

mhevery replied:
fixed

vicb replied:
push ?

mhevery replied:
hmm... not sure what happened, fixed again

mhevery commented on packages/compiler/src/render3/r3_view_compiler.ts:
```diff
@@ -325,34 +331,88 @@ function getLiteralFactory(
+   * - value `lhs` is the left hand side which is an AST representing `abc` which
+   *   can be used in other places of the source code.
+   * - value `rhs` is the right hand side which is an AST representing `a.b.c` which
+   *   can be used in other places of the source code.
```
That was a left over from previous behavior. I have removed it and I think it makes it clearer.

mhevery commented on packages/compiler/src/render3/r3_view_compiler.ts:
```diff
@@ -325,34 +331,88 @@ function getLiteralFactory(
+function noop() {}
+
+/**
+ * Function which is executed whenever a variable is refferenced for the first time in a given
```
fixed

mary-poppins commented:
You can preview f5e88ef at https://pr23168-f5e88ef.ngbuilds.io/.

vicb commented on packages/compiler/src/compiler_util/expression_converter.ts:
```diff
@@ -15,7 +15,46 @@ export class EventHandlerVars { static event = o.variable('$event'); }
+  /**
+   * Store statements which are render3 compatible.
+   */
+  public render3Stmts: o.Statement[];
```
nit: we usually do not specify `public`

mhevery replied:
fixed

vicb commented on packages/compiler/src/compiler_util/expression_converter.ts:
```diff
@@ -15,7 +15,46 @@ export class EventHandlerVars { static event = o.variable('$event'); }
+       */
+      public allowDefault: o.ReadVarExpr) {
+    /**
+     * This is bit of a hack. It converts statments which render2 expects to statements which are
```
And ? TODO ?

mhevery replied:
added todo

vicb commented on packages/compiler/src/render3/r3_view_compiler.ts:
```diff
@@ -325,34 +331,86 @@ function getLiteralFactory(
+  constructor();
+  constructor(parent: BindingScope, declareCallback: DeclareLocalVarCallback);
+  constructor(
+      private parent: BindingScope|null = null,
```
having explicit `null` is to match the TS semantics.
Any good reason to diverge here ?

mhevery replied:
fixed

vicb commented on packages/compiler/src/render3/r3_view_compiler.ts:
```diff
@@ -1052,10 +1112,10 @@ function createHostBindingsFunction(
-    directive: CompileDirectiveMetadata, outputCtx: OutputContext): o.Expression|null {
-  if (Object.getOwnPropertyNames(directive.inputs).length > 0) {
-    return outputCtx.constantPool.getConstLiteral(mapToExpression(directive.inputs));
+function createMapObjectLiteral(
```
Add a one liner comment ? (useful as this does NOT always return a Map Literal despite the name)

mhevery replied:
renamed

vicb commented on packages/compiler/test/render3/r3_view_compiler_template_spec.ts:
```diff
@@ -0,0 +1,102 @@
+    const template = `
+      template:function MyComponent_Template(ctx:any, cm:boolean){
+        if (cm) { 
+          i0.ɵC(0, MyComponent_NgForOf_Template_0, null, _c0); 
```
replace `i0` with `$i0$`

mhevery replied:
done

vicb commented on packages/core/test/bundling/todo/index.ts:
```diff
@@ -6,73 +6,150 @@
-  done: boolean;
+
+export class Todo {
+  completed: Boolean;
```
Boolean -> boolean (replace all)

mhevery replied:
fixed

vicb commented on packages/core/test/bundling/todo/index.ts:
```diff
@@ -6,73 +6,150 @@
-    this.appState = new AppState();
+  // TODO(misko) Fix injection
+  // constructor(todoStore: TodoStore) { this.todoStore = todoStore; }
+  constructor() { this.todoStore = new TodoStore(); }
```
nit: move the declaration ?

vicb reviewed (changes requested):
inline comments

vicb commented on packages/compiler/src/render3/r3_view_compiler.ts:
```diff
@@ -776,9 +839,10 @@ class TemplateDefinitionBuilder implements TemplateAstVisitor, LocalResolver {
     this.instruction(this._creationMode, ast.sourceSpan, R3.text, o.literal(nodeIndex));
 
     // Refresh mode
+    debugger;
```
Oops

mhevery replied:
removed

vicb commented on packages/compiler/src/render3/r3_view_compiler.ts:
```diff
@@ -193,7 +196,10 @@ export function compileComponent(
 
   // e.g `inputs: {a: 'a'}`
-  field('inputs', createInputsObject(component, outputCtx));
+  field('inputs', createMapObjectLiteral(component.inputs, outputCtx));
```
Other fields are defined only when they should (ie `features`, `directives`, `pipes`, ...).
Should we do the same for `inputs` and `outputs` ? 
It would be better for code size.

mhevery replied:
discussed in person, it is OK.

mhevery commented on packages/core/test/bundling/todo/index.ts:
```diff
@@ -6,73 +6,150 @@
+  set title(value: string) { this._title = value.trim(); }
+
+  constructor(title: string, completed: boolean = false) {
+    this.completed = completed;
```
fixe

mhevery commented on packages/core/test/bundling/todo/index.ts:
```diff
@@ -6,73 +6,150 @@
+  constructor(title: string, completed: boolean = false) {
+    this.completed = completed;
+    this.editing = false;
+    this.title = title.trim();
```
removed

mary-poppins commented:
You can preview 2311600 at https://pr23168-2311600.ngbuilds.io/.
